### PR TITLE
Make _id terms optional in segment containing only noop

### PR DESCRIPTION
**Pull request by dnhatn**
Previously only index and delete operations are indexed into Lucene,
therefore every segment should have both _id and _version terms as these
operations contain both terms. However, this is no longer guaranteed
after noop is also indexed into Lucene. A segment which contains only
no-ops does not have neither _id or _version because a no-op does not 
contain these terms.

This change adds a dummy version to no-ops and makes _id terms optional
in PerThreadIDVersionAndSeqNoLookup.

Relates #30226

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 1]
```diff
@@ -3781,6 +3781,51 @@ protected long doGenerateSeqNoForOperation(Operation operation) {
...
+    public void testSegmentContainsOnlyNoOps() throws Exception {
+        Engine.NoOpResult noOpResult = engine.noOp(new Engine.NoOp(1, primaryTerm.get(), REPLICA, randomNonNegativeLong(), "test"));
+        assertThat(noOpResult.getFailure(), nullValue());
+        engine.refresh("test");
+        Engine.DeleteResult deleteResult = engine.delete(replicaDeleteForDoc("id", 1, 2, randomNonNegativeLong()));
+        assertThat(deleteResult.getFailure(), nullValue());
+        engine.refresh("test");
+    }
+
+    public void testRandomOperations() throws Exception {
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: can you add a comment as to why we have this test? something like "a simple test to check that random combination of operations can coexist in segments and be read. This is needed as some fields in lucene may not exist if a segment misses operation types and the code needs to check for that"

File: server/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDVersionAndSeqNoLookup.java [hunk 2]
```diff
@@ -66,15 +66,19 @@
...
-        Terms terms = reader.terms(uidField);
+        final Terms terms = reader.terms(uidField);
         if (terms == null) {
-            throw new IllegalArgumentException("reader misses the [" + uidField + "] field");
-        }
-        termsEnum = terms.iterator();
-        if (reader.getNumericDocValues(VersionFieldMapper.NAME) == null) {
-            throw new IllegalArgumentException("reader misses the [" + VersionFieldMapper.NAME + "] field");
+            // Uid is not found if a segment contains only NoOps; in this case, version should not exist either.
+            if (reader.getNumericDocValues(VersionFieldMapper.NAME) != null) {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDVersionAndSeqNoLookup.java]: I wonder if we can make this more generic and check that the segment only has the noop fields?

File: server/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDVersionAndSeqNoLookup.java [hunk 3]
```diff
@@ -111,7 +115,7 @@ public DocIdAndVersion lookupVersion(BytesRef id, LeafReaderContext context)
      * {@link DocIdSetIterator#NO_MORE_DOCS} is returned if not found
      * */
     private int getDocID(BytesRef id, Bits liveDocs) throws IOException {
-        if (termsEnum.seekExact(id)) {
+        if (termsEnum != null && termsEnum.seekExact(id)) {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDVersionAndSeqNoLookup.java]: can you add a comment as to when the id might not be there?

File: server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 4]
```diff
@@ -3781,6 +3781,51 @@ protected long doGenerateSeqNoForOperation(Operation operation) {
         }
     }
 
+    public void testSegmentContainsOnlyNoOps() throws Exception {
+        Engine.NoOpResult noOpResult = engine.noOp(new Engine.NoOp(1, primaryTerm.get(), REPLICA, randomNonNegativeLong(), "test"));
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: primaries can also have noops upon promotion/recovery. Randomize the engine type?

**Review by bleskes — APPROVED** [review]
LGTM. Good catch. Left some nits.

**Comment by jpountz** [comment]
Change looks good. I'm wondering whether we should add a dummy `_version` to these noop documents, since the doc values impl in the current Lucene codec has better random-access with dense fields. cc @s1monw since we had related discussions in the past.

**Comment by s1monw** [comment]
I was gonna say the same thing as @jpountz that we should make sure we index a dummy version to make the DV dense. I think this is essential for indexing perf.

**Comment by dnhatn** [comment]
@elasticmachine test this please

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 5]
```diff
@@ -1347,6 +1347,9 @@ private NoOpResult innerNoOp(final NoOp noOp) throws IOException {
                 try {
                     final ParsedDocument tombstone = engineConfig.getTombstoneDocSupplier().newNoopTombstoneDoc();
                     tombstone.updateSeqID(noOp.seqNo(), noOp.primaryTerm());
+                    // A noop tombstone does not require a _version but it's added to have a fully dense docvalues for the version field.
+                    // 1L is selected to optimize the compression because it might probably be the most common value in version field.
+                    tombstone.version().setLongValue(1L);
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: this is also the version that append only docs has! 👍

**Review by s1monw — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
Thanks @bleskes @jpountz and @simonw for reviewing!

**Pull request merged**